alvr/common/src/sockets/stream_socket/throttled_udp.rs

```diff
@@ -0,0 +1,149 @@
+use crate::{
+    sockets::{StreamId, LOCAL_IP},
+    *,
+};
+use bytes::{BufMut, Bytes, BytesMut};
+
+use futures::{Stream, StreamExt};
+use governor::{
+    clock,
+    state::{InMemoryState, NotKeyed},
+    Quota, RateLimiter,
+};
+use nonzero_ext::NonZero;
+use std::{
+    collections::HashMap,
+    io,
+    mem::MaybeUninit,
+    net::{IpAddr, SocketAddr},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tokio::{
+    io::ReadBuf,
+    net::UdpSocket,
+    sync::{mpsc, Mutex},
+};
+
+const INITIAL_RD_CAPACITY: usize = 64 * 1024;
+
+// Don't go below this rate, limiting then is unneeded anyway.
+const MINIMUM_BYTERATE: u32 = 30 * 1024 * 1024 * 3 / 2 / 8;
+// Reserve includes audio along with other small fluctuations.
+const RESERVE_BYTERATE: u32 = 5_000_000 / 8;
+
+pub struct ThrottlingSettings {
+    pub byterate: u32,
+    pub multiplier: f32,
+}
+
+#[allow(clippy::type_complexity)]
+#[derive(Clone)]
+pub struct ThrottledUdpStreamSendSocket {
+    inner: Arc<UdpSocket>,
+    limiter: Arc<Option<RateLimiter<NotKeyed, InMemoryState, clock::DefaultClock>>>,
+}
+
+impl ThrottledUdpStreamSendSocket {
+    pub async fn send(&self, data: Bytes) -> io::Result<()> {
+        if let Some(ref limiter) = *self.limiter {
+            if let Some(len) = NonZero::new(data.len() as u32) {
+                limiter.until_n_ready(len).await.ok();
+            }
+        }
+        match self.inner.send(&data).await {
+            Ok(_) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+pub struct ThrottledUdpStreamReceiveSocket {
+    pub inner: Arc<UdpSocket>,
+    buffer: BytesMut,
+}
+
+impl Stream for ThrottledUdpStreamReceiveSocket {
```

Comment:
futures::Stream is quite a low level trait. Are you sure you need this? Personally I would try to avoid implementing traits if not needed by external libraries.